### PR TITLE
Add support for writing certificate signing requests

### PR DESCRIPTION
**Pull request by djc**
Hey, would you be interested in merging something like this? Seems like a feature that mostly fits, at least, although it maybe widens the scope a bit.

On the other hand, I'm not sure how best to test this for now (although I suppose the openssl library exposes API to actually handle CSRs). So far this is based on just comparing openssl-generated CSRs with the JS ASN.1 thingy you point to in your README.

### CODE REVIEW

**Comment by est31** [comment]
@djc thanks for this PR. It's definitely in scope. I've wanted to add this feature eventually myself. It's really cool to have CSR generation because now maybe [acme-client](https://github.com/onur/acme-client) could switch to rcgen.

As for testing I think openssl can be used for it. I suggest creating a `X509Req` and calling its [verify](https://docs.rs/openssl/0.10.23/openssl/x509/struct.X509Req.html#method.verify) function.

**Comment by djc** [comment]
Added a test as per your suggestion.

**Comment by djc** [comment]
(This does introduce the slight naming oddity that a `Certificate` object no longer always represents an actual certificate, because in the CSR case it's merely the `CertificateParams` combined with the certificate's key pair without the means to sign it into a usable certificate. Not sure if that's something you want to address as part of this PR?)

File: src/lib.rs [hunk 1]
```diff
@@ -94,6 +94,9 @@ pub fn generate_simple_self_signed(subject_alt_names :impl Into<Vec<String>>) ->
 // Example certs usable as reference:
 // Uses ECDSA: https://crt.sh/?asn1=607203242
 
+/// pkcs-9-at-extensionRequest in RFC 2985
+const OID_PKCS_9_AT_EXTENSION_REQUEST: &[u64] = &[1, 2, 840, 113549, 1, 9, 14];
```
Comment (est31) [inline src/lib.rs]: code format nit: should be `OID_PKCS_9_AT_EXTENSION_REQUEST :&[u64]`

**Comment by est31** [comment]
> naming oddity that a Certificate object no longer always represents an actual certificate

Yeah I think the API details are out of scope for the PR.

**Comment by djc** [comment]
Also I think at least some versions of ACME require support for custom extensions, which doesn't add yet. But ACME is the motivating use case for me as well, so I can look at supporting that next.

File: src/lib.rs (hunk 1 above)
Comment (djc) [reply src/lib.rs]: Fixed.

**Comment by djc** [comment]
Thanks for the quick review!

**Comment by est31** [comment]
@djc if you want a new release, just ask.

**Comment by djc** [comment]
Will do. I don't think it's needed just yet.

**Comment by djc** [comment]
Do you know if the acme-client maintainer is amenable to replacing openssl with a *ring*-based stack?

**Comment by est31** [comment]
@djc I haven't had any communication with them if that's what you wonder about. I only thought it'd be a nice project.

**Comment by djc** [comment]
The project seems a bit dead. I'm talking to the maintainer of an alternative here: https://github.com/breard-r/acmed/issues/2.

**Comment by est31** [comment]
Oh yeah it kinda is. Getting acmed to use rcgen sounds like a good idea, too.

**Pull request merged**